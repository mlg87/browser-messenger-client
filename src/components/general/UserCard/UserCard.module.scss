.container {
    align-items: center;
    cursor: pointer;
    display: flex;
    padding: 5px 10px;

    &:hover {
        background-color: black;
        color: white;
    }
}

.selectedIndicator {
    color: green;
    margin-right: 5px;
}

.statusIndicator {
    background-color: green;
    border-radius: 100%;
    height: 8px;
    margin-left: 5px;
    width: 8px;
}