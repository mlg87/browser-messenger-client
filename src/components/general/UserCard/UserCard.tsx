import { observer } from 'mobx-react';
import React from 'react';
import { withStores } from '../../../common/utils/withStores';
import { IUser } from '../../../stores/Auth/types';
import { Users } from '../../../stores/Users';
import styles from './UserCard.module.scss';

const UserCard: React.FunctionComponent<IUserCardProps> = props => {

    const { user, users: { selectedUser } } = props;

    const isSelected = selectedUser && selectedUser.id === user.id;

    return (
        <div
            className={styles.container}
            data-testid='user-card'
            onClick={() => _handleClick(props)}
            role='button'
        >
            {isSelected && <div className={styles.selectedIndicator}>>>></div>}
            <div>{user.username}</div>
            {user.isOnline && <div className={styles.statusIndicator} />}
        </div>
    );

};

//
// ─── HANDLERS ───────────────────────────────────────────────────────────────────
//

export const _handleClick = (
    props: IUserCardProps
) => {

    props.users.selectedUser = props.user;

};

export default withStores('users')(observer(UserCard));

//
// ─── INTERFACES ─────────────────────────────────────────────────────────────────
//

export interface IUserCardProps {
    readonly user: IUser;
    readonly users: Users;
}
