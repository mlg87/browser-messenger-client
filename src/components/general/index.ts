export { default as ChatBox } from './ChatBox/ChatBox';
export { default as UserCard } from './UserCard/UserCard';
export { default as UserList } from './UserList/UserList';
