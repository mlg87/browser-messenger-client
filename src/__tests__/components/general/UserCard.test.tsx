import { act, fireEvent, render } from '@testing-library/react';
import React from 'react';
import { UserCard } from '../../../components/general';
import { IUserCardProps } from '../../../components/general/UserCard/UserCard';
import appState from '../../../stores';

const startingProps: IUserCardProps = {
    user: {
        id: 123,
        isOnline: true,
        username: 'test'
    },
    users: appState.users
};

const setup = (props: IUserCardProps = startingProps) => render(
    <UserCard  {...props} />
);

describe('UserCard', () => {

    it('renders', () => {

        const { container } = setup();
        expect(container.firstChild).toMatchSnapshot();

    });

    it('sets the selected user when the user clicks the card', () => {

        const { container, getByRole } = setup();
        const card = getByRole('button');

        act(() => {

            fireEvent.click(card);

        });

        expect(container.firstChild).toMatchSnapshot();
        expect(appState.users.selectedUser).toBeDefined();
        expect(appState.users.selectedUser!.id).toEqual(123);

    });

});
