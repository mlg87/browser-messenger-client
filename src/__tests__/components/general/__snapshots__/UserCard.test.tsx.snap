// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCard renders 1`] = `
<div
  class="container"
  data-testid="user-card"
  role="button"
>
  <div>
    test
  </div>
  <div
    class="statusIndicator"
  />
</div>
`;

exports[`UserCard sets the selected user when the user clicks the card 1`] = `
<div
  class="container"
  data-testid="user-card"
  role="button"
>
  <div
    class="selectedIndicator"
  >
    &gt;&gt;&gt;
  </div>
  <div>
    test
  </div>
  <div
    class="statusIndicator"
  />
</div>
`;
